### ISSUE TITLE
_broadcast_shape is broken when 32 < nargs

### ISSUE THREAD

**Issue by behzadnouri** [numpy/numpy#5862]
xref https://github.com/numpy/numpy/pull/5371

```python
>>> np.__version__
'1.10.0.dev0+30e3d41'
>>> from numpy.lib.stride_tricks import _broadcast_shape
>>> lst = [np.ones((5, 7, 11))] * 32 + [np.ones((2, 3))] * 32
>>> _broadcast_shape(*lst)
(2, 3)
>>> _broadcast_shape(*lst[::-1])  # process the list in reverse order
(5, 7, 11)
```

In [this line](https://github.com/numpy/numpy/blob/05b5335ecf25e59477956b4f85b9a8edbdf71bcc/numpy/lib/stride_tricks.py#L123) `b` is ultimately converted to an array of shape `1 x 1`; so it will broadcast with the rest of args regardless of their shape.

**Comment by shoyer** [comment]
Thanks for pointing this out. The issue appears to be that `np.broadcast` objects does not in fact handle `np.broadcast` arguments properly -- instead they are treated as scalar arrays (@@:
```
In [22]: np.broadcast(np.broadcast([1], [2]), 1).shape
Out[22]: ()
```

To be clear, it also means that the current implementation of `np.broadcast_arrays` (on master) is buggy. It would be good to fix this prior to the 1.10 release. I'll submit a PR with a fix shortly.

CC @mhvk

**Comment by mhvk** [comment]
@behzadnouri - unexpected indeed, and thanks for tracing it down to the `np.broadcast` line; @shoyer - bizarre indeed that `np.broadcast` cannot handle its own objects! PR looks good, but is it worth raising a separate issue about `np.broadcast` not handling its own output?
